### PR TITLE
Fix DAG Generation and Enable Handling Initial Conditions

### PR DESCRIPTION
**Pull request by mjs271**
This PR modifies the atmospheric DAG generator (`atmosphere_process_dag.xpp`) and atmosphere driver to get things working again. Namely,

- There are now 4 checkpoints during `AtmosphereDriver::initialize()` at which we build the most complete DAG available and write it to file.
    - This is intended to serve as a diagnostic tool in case a build crashes, giving the user information about the created/initialized processes or fields.
    - Some examples are below that show the incremental increase in the displayed information.
    - These checkpoints occur at the end of the following functions, and the associated DAG is written out to a `.dot` file in the run directory:
          1. `create_atm_processes()`--scream_atm_createProc_dag.dot
          2. `create_fields()`--scream_atm_createField_dag.dot
          3. `initialize_fields()`--scream_atm_initField_dag.dot
          4. `initialize_atm_procs()`--scream_atm_initProc_dag.dot
- I've added some code that enables the DAG to properly handle initial conditions and determine whether they are missing at the time when he DAG is generated.
    - If, at the time of generation, a `required` field (or `group`) is missing it is indicated on the corresponding node, as can be seen in the 2nd or 3rd image below.
    - Before the 4th and final DAG, (within `initialize_atm_procs()`), an extra box is added to the DAG with a disclaimer that any fields indicated as missing may be provided by the forthcoming initial condition file.
    - A node for the Initial Condition is now displayed on the DAG with edges leading to the
- I've made some minor formatting changes for readability and to more clearly indicate the different types of objects represented on the DAG--e.g., the **Begin/End of Atm. Timestep** nodes are colored differently, and fields are printed in green when provided directly by the initial condition at initialization.
- I've also fenced-off the `write_dag()` statements to ensure they are only printed by the main MPI thread, since I noticed occasional formatting errors due to a presumptive race condition.

createField_dag
initField_dag
initProc_dag

### CODE REVIEW

**Comment by bartgol** [comment]
Before reviewing, a couple of comments:
- what's the point of the 1st checkpoint? It seems to only print boxes for processes, without edges. I think we can get rid of that (unless I am missing something).
- I don't see what fig 3 adds. It's an interpolation point between 2 and 4. I think we can do away with it. Maybe we can add a box message to fig 4 that says "green fields are read from IC file and never updated", if you really want to be verbose.

Bottom line, I think we can reduce to 2 dags: before and after reading the IC file.

**Comment by mjs271** [comment]
> what's the point of the 1st checkpoint? It seems to only print boxes for processes, without edges. I think we can get rid of that (unless I am missing something).

I'm good with getting rid of if. My thought was to generate the DAG each time, during initialization, that more information is available that'd be reflected in the DAG, in case the driver crashes immediately after the point that the DAG is generated. For the first checkpoint, boxes with no edges, it at least reports what processes have been successfully initialized.

However, I do agree that it's a low-information figure, so I'm good to lose it--unless generating it and then overwriting with the next DAG iteration is preferred (the previous behavior of the DAG workflow).

> I don't see what fig 3 adds. It's an interpolation point between 2 and 4. I think we can do away with it. Maybe we can add a box message to fig 4 that says "green fields are read from IC file and never updated", if you really want to be verbose.

Also good with losing fig 3 based on the above perspective.

I'll add that info box--do you have strong opinions on whether you'd like to control that behavior with verbosity level (more so than it already is)? @bartgol

File: components/eamxx/src/control/atmosphere_driver.cpp [hunk 1]
```diff
@@ -685,6 +699,23 @@ void AtmosphereDriver::create_fields()
...
+  auto& driver_options_pl = m_atm_params.sublist("driver_options");
+  const int verb_lvl = driver_options_pl.get<int>("atmosphere_dag_verbosity_level",-1);
+  if (verb_lvl>0) {
+    // now that we've got fields, generate a DAG with fields and dependencies
+    // NOTE: at this point, fields provided by initial conditions may (will)
+    // appear as unmet dependencies
+    AtmProcDAG dag;
+
+    // First, add all atm processes
+    dag.create_dag(*m_atm_process_group);
```
Comment (tcclevenger) [inline components/eamxx/src/control/atmosphere_driver.cpp:711]: I'm not too familiar with the DAG, but would it make sense to have a class variable for the driver
```
std::shared_ptr<AtmProcDAG> m_dag;
```
and then whenever we have the `m_atm_process_group` setup, call something like 
```
if (verb_lvl>0) {
  auto m_dag = std::make_shared<AtmProcDAG>();
  m_dag->create_dag(*m_atm_process_group);
}
```
Then in all these places we only need to call `m_dag->write_dag(...)`.

File: components/eamxx/src/control/atmosphere_driver.cpp [hunk 2]
```diff
@@ -1143,7 +1176,7 @@ void AtmosphereDriver::set_initial_conditions ()
           } else {
             this_grid_ic_fnames.push_back(fname);
           }
-	  fields_inited[grid_name].push_back(fname);
+          m_fields_inited[grid_name].push_back(fname);
```
Comment (tcclevenger) [inline components/eamxx/src/control/atmosphere_driver.cpp:1179]: Maybe then at the end of this block call
```
m_dag-> process_initial_conditions(m_fields_inited);
```

**Comment by bartgol** [comment]
@mjs271 I think the real purpose of the DAG is to see what the dependencies are if something fails at runtime (and possibly to throw if deps are unmet during init). If the code crashes for whatever other reason during init, I doubt that inspecting the DAG would offer any insight. 

If the DAG is needed for debug, we're likely past the point where they are all written, and since the user can ask for more/less detail with the dag verbosity param, I think we're covered there.

> I'll add that info box--do you have strong opinions on whether you'd like to control that behavior with verbosity level (more so than it already is)? @bartgol
I think you can put that box in as soon as the verb level is such that fields names are written in the nodes

File: components/eamxx/src/control/atmosphere_driver.cpp (hunk 1 above)
Comment (mjs271) [reply components/eamxx/src/control/atmosphere_driver.cpp:711]: I like this idea in principal, however I do think it would create a small headache spot with regard to how initialization happens with the PG2 grid (this is partly discussed in #3018). I haven't chased it down to the exact workings, but at some point, the PG2 fields are considered missing and may or may not be denoted as coming from the GLL grid. This does not appear to fully resolve until `initialize_atm_procs()` is called, but changes do appear to occur when `initialize_fields()` is called. I've got some example DAGs below to demonstrate this as it happens in the `homme_shoc_cld_spa_p3_rrtmgp_pg2_dp` test.

So, my thought is that regenerating and destroying, while non-optimal, is the simpler option because it's conducted  during initialization and should be relatively low in memory and cycles. I'm certainly willing to be persuaded otherwise, though--will just require a little retooling.

scream_atm_createField_dag
scream_atm_initField_dag
scream_atm_initProc_dag

**Comment by bartgol** [comment]
@mjs271 What's the status? Are you planning on pushing new changes, or are you waiting on feedback?

**Comment by mjs271** [comment]
@bartgol Just got back to this and have some changes pushed. I'm happy with it if others are

**Comment by mjs271** [comment]
Example of newly-added explanatory box:

scream_atm_initProc_dag

**Review by bartgol — DISMISSED** [review]

**Review by tcclevenger — DISMISSED** [review]

**Comment by AaronDonahue** [comment]
Will this address #1256 ?  @luca?

**Comment by bartgol** [comment]
> Will this address #1256 ? @luca?

I think you got the wrong Luca there. :)

I think that issue is broader, but I also think we can close it. I don't think we want the IC to be its own atm process, considering that it would only run during init.

**Comment by mjs271** [comment]
@bartgol --in response to your [comment](https://github.com/E3SM-Project/scream/issues/3018#issuecomment-2477856986), maybe best to merge this PR now/soon to avoid additional delay? Then I'll open a new one that addresses #3018, as well as:
- runtime checks for missing/uninitialized fields and erroring out on failure
- adding a `WILL_FAIL` test that confirms detection of missing/uninitialized fields

**Comment by bartgol** [comment]
@mjs271 We do get some SIGSEV error in the CUDA builds. Might be worth taking a look.

**Comment by mjs271** [comment]
@bartgol haha, thanks--rebased and reran to see if that could be the issue.  Will knock that out shortly! 😅
